scripts/core/with-live-query.tsx

```diff
@@ -0,0 +1,148 @@
+
+type IState<T extends IBaseRestApiResponse> = {loading: true} | IStateReady<T>;
+
+interface IStateReady<T extends IBaseRestApiResponse> {
+    data: IRestApiResponse<T>;
+    loading: false;
+}
+
+export class WithLiveQuery<T extends IBaseRestApiResponse> extends React.PureComponent<ILiveQueryProps<T>, IState<T>> {
+    private eventListenersToRemoveBeforeUnmounting: Array<() => void>;
+    private handleContentChangesThrottled: (changes: Array<IResourceChange>) => void;
+
+    constructor(props: ILiveQueryProps<T>) {
+        super(props);
+
+        this.state = {loading: true};
+
+        this.eventListenersToRemoveBeforeUnmounting = [];
+
+        this.fetchItems = this.fetchItems.bind(this);
+        this.handleContentChanges = this.handleContentChanges.bind(this);
+
+        this.handleContentChangesThrottled = throttleAndCombineArray(
+            (changes: Array<IResourceChange>) => {
+                this.handleContentChanges(changes);
+            },
+            300,
+        );
+
+        this.eventListenersToRemoveBeforeUnmounting.push(
+            addWebsocketEventListener(
+                'resource:created',
+                (event) => {
+                    const {resource, _id} = event.extra;
+
+                    this.handleContentChangesThrottled([{changeType: 'created', resource: resource, itemId: _id}]);
+                },
+            ),
+        );
+
+        this.eventListenersToRemoveBeforeUnmounting.push(
+            addWebsocketEventListener(
+                'resource:updated',
+                (event) => {
+                    const {resource, _id, fields} = event.extra;
+
+                    this.handleContentChangesThrottled([{
+                        changeType: 'updated',
+                        resource: resource,
+                        itemId: _id,
+                        fields: fields,
+                    }]);
+                },
+            ),
+        );
+
+        this.eventListenersToRemoveBeforeUnmounting.push(
+            addWebsocketEventListener(
+                'resource:deleted',
+                (event) => {
+                    const {resource, _id} = event.extra;
+
+                    this.handleContentChangesThrottled([{changeType: 'deleted', resource: resource, itemId: _id}]);
+                },
+            ),
+        );
+    }
+
+    fetchItems() {
+        const {resource, query} = this.props;
+
+        return httpRequestJsonLocal<IRestApiResponse<T>>(
+            {
+                method: 'GET',
+                path: '/' + resource,
+                urlParams: {
+                    aggregations: 0,
+                    es_highlight: 1,
+                    source: JSON.stringify(toElasticQuery(query)),
+                },
+            },
+        ).then((data) => {
+            this.setState({data: data, loading: false});
+        });
+    }
+
+    handleContentChanges(changes: Array<IResourceChange>) {
+        if (this.state.loading === true) {
```

Comment:
if I get it right it will ignore any changes done while it's loading data right?

Reply:
Yes, and that is a problem actually, because we would end up with old data. Instead I should probably simply call the throttled version so it queues another update.

Reply:
but when do you set loading to true again? it seems like that's only set in constructor so all further updates should be fetching from the server

Reply:
Good catch. I'll set it on every request.